vm/src/stdlib/socket.rs

```diff
@@ -352,6 +362,85 @@ impl SocketRef {
             Err(s) => Err(vm.new_os_error(s.to_string())),
         }
     }
+
+    fn gettimeout(self, vm: &VirtualMachine) -> PyResult {
+        match self.timeout.borrow().as_ref() {
+            Some(duration) => Ok(vm.ctx.new_float(duration.as_secs() as f64)),
+            None => Ok(vm.get_none()),
+        }
+    }
+
+    fn setblocking(self, block: PyObjectRef, vm: &VirtualMachine) -> PyResult<()> {
+        match boolval(vm, block.clone()) {
+            Ok(value) => {
+                if value {
+                    self.timeout.replace(None);
+                } else {
+                    self.timeout.borrow_mut().replace(Duration::from_secs(0));
+                }
+                if self.internal_setblocking(value) {
+                    Ok(())
+                } else {
+                    Err(vm.new_os_error("Set socket blocking error".to_string()))
+                }
+            }
+            Err(_) => Err(vm.new_type_error(format!(
+                "an integer os required (got type {})",
+                &block.class().name.clone()
+            ))),
+        }
+    }
+
+    fn internal_setblocking(&self, block: bool) -> bool {
+        match self.con.borrow_mut().as_mut() {
+            Some(conn) => match conn {
+                Connection::TcpStream(con) => con.set_nonblocking(!block).is_ok(),
+                Connection::TcpListener(con) => con.set_nonblocking(!block).is_ok(),
+                Connection::UdpSocket(con) => con.set_nonblocking(!block).is_ok(),
+            },
+            None => true,
+        }
+    }
+
+    fn getblocking(self, vm: &VirtualMachine) -> PyResult {
+        Ok(vm.ctx.new_bool(self.internal_getblocking()))
+    }
+
+    fn internal_getblocking(&self) -> bool {
+        match self.timeout.borrow().as_ref() {
+            Some(duration) => {
+                if duration.as_secs() != 0 {
+                    true
+                } else {
+                    false
+                }
+            }
+            None => true,
+        }
+    }
+
+    fn settimeout(self, timeout: PyObjectRef, vm: &VirtualMachine) -> PyResult<()> {
```

Comment:
I met some trouble when I try to implement the settimeout. If I set timeout, Rust will throw a Resource unavailable error instead of timeout error when read timeout, I don't know if I should handle this error.  Do you have any good suggestions?

Reply:
Does this happens only on `TcpListener`? I see that the others were implemented

Reply:
TcpListener doesn't have a method to set timeout, So I haven’t implemented it yet.
I try to [catch the timeout error](https://github.com/RustPython/RustPython/pull/1309/files#diff-eeb846e2bb3ddeef931431d5a62de4ecR251) in recv, But after the timeout, what is thrown is the Resource temporarily unavailable error. There are many reasons for this error, so I don't know if it should be converted into a timeout error.

Reply:
We can work on that later. I think we should merge this as it is.